File: server/src/websocket.rs
use axum::{
    extract::{
        ws::{Message, WebSocket, WebSocketUpgrade},
        ConnectInfo, State,
    },
    response::IntoResponse,
};
use chrono::Utc;
use futures::{SinkExt, StreamExt};
use std::net::SocketAddr;
use tokio::sync::mpsc;

use crate::db::store_metrics;
use crate::state::AppState;
use crate::types::{AgentMessage, AgentMetricsData, DashboardMessage, ServerMetricsUpdate};

// ============================================================================
// Dashboard WebSocket Handler
// ============================================================================

pub async fn ws_handler(ws: WebSocketUpgrade, State(state): State<AppState>) -> impl IntoResponse {
    ws.on_upgrade(move |socket| handle_dashboard_socket(socket, state))
}

async fn handle_dashboard_socket(socket: WebSocket, state: AppState) {
    let (mut sender, mut receiver) = socket.split();
    let mut rx = state.metrics_tx.subscribe();

    // Send initial state with site settings
    {
        let config = state.config.read().await;
        let agent_metrics = state.agent_metrics.read().await;

        let updates: Vec<ServerMetricsUpdate> = config
            .servers
            .iter()
            .map(|server| {
                let metrics_data = agent_metrics.get(&server.id);
                let online = metrics_data
                    .map(|m| Utc::now().signed_duration_since(m.last_updated).num_seconds() < 30)
                    .unwrap_or(false);

                let version = metrics_data
                    .and_then(|m| m.metrics.version.clone())
                    .unwrap_or_else(|| server.version.clone());

                ServerMetricsUpdate {
                    server_id: server.id.clone(),
                    server_name: server.name.clone(),
                    location: server.location.clone(),
                    provider: server.provider.clone(),
                    tag: server.tag.clone(),
                    version,
                    ip: server.ip.clone(),
                    online,
                    metrics: metrics_data.map(|m| m.metrics.clone()),
                }
            })
            .collect();

        let msg = DashboardMessage {
            msg_type: "metrics".to_string(),
            servers: updates,
            site_settings: Some(config.site_settings.clone()),
        };

        if let Ok(json) = serde_json::to_string(&msg) {
            let _ = sender.send(Message::Text(json.into())).await;
        }
    }

    let send_task = tokio::spawn(async move {
        while let Ok(msg) = rx.recv().await {
            if sender.send(Message::Text(msg.into())).await.is_err() {
                break;
            }
        }
    });

    let recv_task = tokio::spawn(async move {
        while let Some(Ok(msg)) = receiver.next().await {
            if let Message::Close(_) = msg {
                break;
            }
        }
    });

    tokio::select! {
        _ = send_task => {},
        _ = recv_task => {},
    }
}

// ============================================================================
// Agent WebSocket Handler
// ============================================================================

pub async fn agent_ws_handler(
    ws: WebSocketUpgrade,
    State(state): State<AppState>,
    ConnectInfo(addr): ConnectInfo<SocketAddr>,
) -> impl IntoResponse {
    let client_ip = addr.ip().to_string();
    ws.on_upgrade(move |socket| handle_agent_socket(socket, state, client_ip))
}

async fn handle_agent_socket(socket: WebSocket, state: AppState, client_ip: String) {
    let (mut sender, mut receiver) = socket.split();
    let mut authenticated_server_id: Option<String> = None;
    
    // Create channel for sending commands to this agent
    let (cmd_tx, mut cmd_rx) = mpsc::channel::<Message>(16);
    
    tracing::debug!("Agent connection from IP: {}", client_ip);

    loop {
        tokio::select! {
            // Handle incoming messages from agent
            msg = receiver.next() => {
                match msg {
                    Some(Ok(Message::Text(text))) => {
                        if let Ok(agent_msg) = serde_json::from_str::<AgentMessage>(&text) {
                            match agent_msg.msg_type.as_str() {
                                "auth" => {
                                    if let (Some(server_id), Some(token)) =
                                        (agent_msg.server_id, agent_msg.token)
                                    {
                                        let config = state.config.read().await;
                                        if let Some(server) =
                                            config.servers.iter().find(|s| s.id == server_id)
                                        {
                                            if server.token == token {
                                                authenticated_server_id = Some(server_id.clone());
                                                
                                                // Register this agent's command channel
                                                {
                                                    let mut connections = state.agent_connections.write().await;
                                                    connections.insert(server_id.clone(), cmd_tx.clone());
                                                }
                                                
                                                let _ = sender
                                                    .send(Message::Text(
                                                        r#"{"type":"auth","status":"ok"}"#.into(),
                                                    ))
                                                    .await;
                                                tracing::info!("Agent {} authenticated and registered", server_id);
                                            } else {
                                                let _ = sender
                                                    .send(Message::Text(
                                                        r#"{"type":"auth","status":"error","message":"Invalid token"}"#
                                                            .into(),
                                                    ))
                                                    .await;
                                            }
                                        } else {
                                            let _ = sender
                                                .send(Message::Text(
                                                    r#"{"type":"auth","status":"error","message":"Server not found"}"#
                                                        .into(),
                                                ))
                                                .await;
                                        }
                                    }
                                }
                                "metrics" => {
                                    if let Some(ref server_id) = authenticated_server_id {
                                        if let Some(metrics) = agent_msg.metrics {
                                            // Store to database
                                            {
                                                let db = state.db.lock().await;
                                                if let Err(e) = store_metrics(&db, server_id, &metrics) {
                                                    tracing::warn!("Failed to store metrics: {}", e);
                                                }
                                            }

                                            // Determine the IP address to use:
                                            // 1. Use agent-reported IPs if available
                                            // 2. Fall back to client connection IP
                                            let agent_ip = metrics.ip_addresses
                                                .as_ref()
                                                .and_then(|ips| ips.first().cloned())
                                                .unwrap_or_else(|| client_ip.clone());

                                            // Update version and IP in server config if provided
                                            {
                                                let mut config = state.config.write().await;
                                                if let Some(server) = config.servers.iter_mut().find(|s| s.id == *server_id) {
                                                    let mut changed = false;
                                                    
                                                    if let Some(ref version) = metrics.version {
                                                        if server.version != *version {
                                                            server.version = version.clone();
                                                            changed = true;
                                                        }
                                                    }
                                                    
                                                    // Update IP if changed
                                                    if server.ip != agent_ip {
                                                        server.ip = agent_ip.clone();
                                                        changed = true;
                                                        tracing::info!("Agent {} IP updated to: {}", server_id, agent_ip);
                                                    }
                                                    
                                                    if changed {
                                                        crate::config::save_config(&config);
                                                    }
                                                }
                                            }

                                            // Update in-memory state
                                            let mut agent_metrics = state.agent_metrics.write().await;
                                            agent_metrics.insert(
                                                server_id.clone(),
                                                AgentMetricsData {
                                                    server_id: server_id.clone(),
                                                    metrics: metrics.clone(),
                                                    last_updated: Utc::now(),
                                                },
                                            );

                                            // Broadcast to dashboard clients
                                            let config = state.config.read().await;
                                            let updates: Vec<ServerMetricsUpdate> = config
                                                .servers
                                                .iter()
                                                .map(|server| {
                                                    let metrics_data = agent_metrics.get(&server.id);
                                                    let online = metrics_data
                                                        .map(|m| {
                                                            Utc::now()
                                                                .signed_duration_since(m.last_updated)
                                                                .num_seconds()
                                                                < 30
                                                        })
                                                        .unwrap_or(false);

                                                    let version = metrics_data
                                                        .and_then(|m| m.metrics.version.clone())
                                                        .unwrap_or_else(|| server.version.clone());

                                                    ServerMetricsUpdate {
                                                        server_id: server.id.clone(),
                                                        server_name: server.name.clone(),
                                                        location: server.location.clone(),
                                                        provider: server.provider.clone(),
                                                        tag: server.tag.clone(),
                                                        version,
                                                        ip: server.ip.clone(),
                                                        online,
                                                        metrics: metrics_data.map(|m| m.metrics.clone()),
                                                    }
                                                })
                                                .collect();

                                            let msg = DashboardMessage {
                                                msg_type: "metrics".to_string(),
                                                servers: updates,
                                                site_settings: None,
                                            };

                                            if let Ok(json) = serde_json::to_string(&msg) {
                                                let _ = state.metrics_tx.send(json);
                                            }
                                        }
                                    } else {
                                        let _ = sender
                                            .send(Message::Text(
                                                r#"{"type":"error","message":"Not authenticated"}"#.into(),
                                            ))
                                            .await;
                                    }
                                }
                                _ => {}
                            }
                        }
                    }
                    Some(Ok(Message::Close(_))) => break,
                    Some(Err(_)) | None => break,
                    _ => {}
                }
            }
            
            // Handle commands from server to agent
            cmd = cmd_rx.recv() => {
                match cmd {
                    Some(message) => {
                        if sender.send(message).await.is_err() {
                            break;
                        }
                    }
                    None => break,
                }
            }
        }
    }

    // Cleanup on disconnect
    if let Some(server_id) = authenticated_server_id {
        tracing::info!("Agent {} disconnected", server_id);
        
        // Remove from active connections
        {
            let mut connections = state.agent_connections.write().await;
            connections.remove(&server_id);
        }
        
        let config = state.config.read().await;
        let agent_metrics = state.agent_metrics.read().await;

        let updates: Vec<ServerMetricsUpdate> = config
            .servers
            .iter()
            .map(|server| {
                let metrics_data = agent_metrics.get(&server.id);
                let online = if server.id == server_id {
                    false
                } else {
                    metrics_data
                        .map(|m| {
                            Utc::now()
                                .signed_duration_since(m.last_updated)
                                .num_seconds()
                                < 30
                        })
                        .unwrap_or(false)
                };

                let version = metrics_data
                    .and_then(|m| m.metrics.version.clone())
                    .unwrap_or_else(|| server.version.clone());

                ServerMetricsUpdate {
                    server_id: server.id.clone(),
                    server_name: server.name.clone(),
                    location: server.location.clone(),
                    provider: server.provider.clone(),
                    tag: server.tag.clone(),
                    version,
                    ip: server.ip.clone(),
                    online,
                    metrics: metrics_data.map(|m| m.metrics.clone()),
                }
            })
            .collect();

        let msg = DashboardMessage {
            msg_type: "metrics".to_string(),
            servers: updates,
            site_settings: None,
        };

        if let Ok(json) = serde_json::to_string(&msg) {
            let _ = state.metrics_tx.send(json);
        }
    }
}

File: server-go/websocket.go
package main

import (
	"encoding/json"
	"log"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/gorilla/websocket"
)

var upgrader = websocket.Upgrader{
	CheckOrigin: func(r *http.Request) bool {
		return true
	},
}

// ============================================================================
// Dashboard WebSocket Handler
// ============================================================================

func (s *AppState) HandleDashboardWS(c *gin.Context) {
	conn, err := upgrader.Upgrade(c.Writer, c.Request, nil)
	if err != nil {
		log.Printf("WebSocket upgrade error: %v", err)
		return
	}
	defer conn.Close()

	// Get client IP
	clientIP := c.ClientIP()

	// Register client with IP
	s.DashboardMu.Lock()
	s.DashboardClients[conn] = &DashboardClient{
		Conn: conn,
		IP:   clientIP,
	}
	s.DashboardMu.Unlock()

	// Unregister on exit
	defer func() {
		s.DashboardMu.Lock()
		delete(s.DashboardClients, conn)
		s.DashboardMu.Unlock()
	}()

	// Send initial state
	s.sendInitialState(conn)

	// Handle incoming messages
	for {
		_, _, err := conn.ReadMessage()
		if err != nil {
			break
		}
	}
}

func (s *AppState) sendInitialState(conn *websocket.Conn) {
	s.ConfigMu.RLock()
	config := s.Config
	s.ConfigMu.RUnlock()

	s.AgentMetricsMu.RLock()
	agentMetrics := make(map[string]*AgentMetricsData)
	for k, v := range s.AgentMetrics {
		agentMetrics[k] = v
	}
	s.AgentMetricsMu.RUnlock()

	// Collect local metrics
	localMetrics := CollectMetrics()

	var updates []ServerMetricsUpdate

	// Add local node
	localNode := config.LocalNode
	localName := "Dashboard Server"
	if localNode.Name != "" {
		localName = localNode.Name
	}
	provider := "Local"
	if localNode.Provider != "" {
		provider = localNode.Provider
	}

	updates = append(updates, ServerMetricsUpdate{
		ServerID:     "local",
		ServerName:   localName,
		Location:     localNode.Location,
		Provider:     provider,
		Tag:          localNode.Tag,
		GroupID:      localNode.GroupID,
		GroupValues:  localNode.GroupValues,
		Version:      ServerVersion,
		IP:           "",
		Online:       true,
		Metrics:      &localMetrics,
		PriceAmount:  localNode.PriceAmount,
		PricePeriod:  localNode.PricePeriod,
		PurchaseDate: localNode.PurchaseDate,
		TipBadge:     localNode.TipBadge,
	})

	// Add remote servers
	for _, server := range config.Servers {
		metricsData := agentMetrics[server.ID]
		online := false
		if metricsData != nil {
			online = time.Since(metricsData.LastUpdated).Seconds() < 30
		}

		version := server.Version
		if metricsData != nil && metricsData.Metrics.Version != "" {
			version = metricsData.Metrics.Version
		}

		var metrics *SystemMetrics
		if metricsData != nil {
			metrics = &metricsData.Metrics
		}

		updates = append(updates, ServerMetricsUpdate{
			ServerID:     server.ID,
			ServerName:   server.Name,
			Location:     server.Location,
			Provider:     server.Provider,
			Tag:          server.Tag,
			GroupID:      server.GroupID,
			GroupValues:  server.GroupValues,
			Version:      version,
			IP:           server.IP,
			Online:       online,
			Metrics:      metrics,
			PriceAmount:  server.PriceAmount,
			PricePeriod:  server.PricePeriod,
			PurchaseDate: server.PurchaseDate,
			TipBadge:     server.TipBadge,
		})
	}

	msg := DashboardMessage{
		Type:            "metrics",
		Servers:         updates,
		Groups:          config.Groups,
		GroupDimensions: config.GroupDimensions,
		SiteSettings:    &config.SiteSettings,
	}

	data, _ := json.Marshal(msg)
	conn.WriteMessage(websocket.TextMessage, data)
}

func (s *AppState) BroadcastMetrics(msg string) {
	s.DashboardMu.RLock()
	defer s.DashboardMu.RUnlock()

	for conn, client := range s.DashboardClients {
		if client != nil && client.Conn != nil {
			if err := conn.WriteMessage(websocket.TextMessage, []byte(msg)); err != nil {
				delete(s.DashboardClients, conn)
				conn.Close()
			}
		}
	}
}

// ============================================================================
// Agent WebSocket Handler
// ============================================================================

func (s *AppState) HandleAgentWS(c *gin.Context) {
	conn, err := upgrader.Upgrade(c.Writer, c.Request, nil)
	if err != nil {
		log.Printf("WebSocket upgrade error: %v", err)
		return
	}
	defer conn.Close()

	clientIP := c.ClientIP()
	var authenticatedServerID string

	// Create channel for sending commands
	sendChan := make(chan []byte, 16)
	done := make(chan struct{})

	// Goroutine to send commands to agent
	go func() {
		for {
			select {
			case msg := <-sendChan:
				if err := conn.WriteMessage(websocket.TextMessage, msg); err != nil {
					log.Printf("Failed to send message to agent: %v", err)
					return
				}
			case <-done:
				return
			}
		}
	}()

	// Handle incoming messages
	for {
		_, message, err := conn.ReadMessage()
		if err != nil {
			break
		}

		var agentMsg AgentMessage
		if err := json.Unmarshal(message, &agentMsg); err != nil {
			continue
		}

		switch agentMsg.Type {
		case "auth":
			if agentMsg.ServerID != "" && agentMsg.Token != "" {
				s.ConfigMu.Lock()
				var server *RemoteServer
				for i := range s.Config.Servers {
					if s.Config.Servers[i].ID == agentMsg.ServerID {
						if s.Config.Servers[i].Token == agentMsg.Token {
							server = &s.Config.Servers[i]
							authenticatedServerID = agentMsg.ServerID

							// Update version
							if agentMsg.Version != "" && server.Version != agentMsg.Version {
								server.Version = agentMsg.Version
								SaveConfig(s.Config)
							}

							// Register connection
							s.AgentConnsMu.Lock()
							s.AgentConns[agentMsg.ServerID] = &AgentConnection{
								Conn:     conn,
								SendChan: sendChan,
							}
							s.AgentConnsMu.Unlock()

							// Send auth success with probe config
							response := map[string]interface{}{
								"type":  "auth",
								"status": "ok",
							}
							if len(s.Config.ProbeSettings.PingTargets) > 0 {
								response["ping_targets"] = s.Config.ProbeSettings.PingTargets
							}
							data, _ := json.Marshal(response)
							conn.WriteMessage(websocket.TextMessage, data)
							log.Printf("Agent %s authenticated", agentMsg.ServerID)
						} else {
							conn.WriteMessage(websocket.TextMessage, []byte(`{"type":"auth","status":"error","message":"Invalid token"}`))
						}
						break
					}
				}
				if server == nil {
					conn.WriteMessage(websocket.TextMessage, []byte(`{"type":"auth","status":"error","message":"Server not found"}`))
				}
				s.ConfigMu.Unlock()
			}

		case "metrics":
			if authenticatedServerID != "" && agentMsg.Metrics != nil {
				// Store to database
				if s.DB != nil {
					go StoreMetrics(s.DB, authenticatedServerID, agentMsg.Metrics)
				}

				// Determine IP address
				agentIP := clientIP
				if len(agentMsg.Metrics.IPAddresses) > 0 {
					agentIP = agentMsg.Metrics.IPAddresses[0]
				}

				// Update version and IP in config
				s.ConfigMu.Lock()
				for i := range s.Config.Servers {
					if s.Config.Servers[i].ID == authenticatedServerID {
						changed := false
						if agentMsg.Metrics.Version != "" && s.Config.Servers[i].Version != agentMsg.Metrics.Version {
							s.Config.Servers[i].Version = agentMsg.Metrics.Version
							changed = true
						}
						if s.Config.Servers[i].IP != agentIP {
							s.Config.Servers[i].IP = agentIP
							changed = true
						}
						if changed {
							SaveConfig(s.Config)
						}
						break
					}
				}
				s.ConfigMu.Unlock()

				// Update in-memory state
				s.AgentMetricsMu.Lock()
				s.AgentMetrics[authenticatedServerID] = &AgentMetricsData{
					ServerID:    authenticatedServerID,
					Metrics:     *agentMsg.Metrics,
					LastUpdated: time.Now(),
				}
				s.AgentMetricsMu.Unlock()
			} else {
				conn.WriteMessage(websocket.TextMessage, []byte(`{"type":"error","message":"Not authenticated"}`))
			}
		}
	}

	// Cleanup on disconnect
	close(done) // Stop the send goroutine
	if authenticatedServerID != "" {
		log.Printf("Agent %s disconnected", authenticatedServerID)
		s.AgentConnsMu.Lock()
		delete(s.AgentConns, authenticatedServerID)
		s.AgentConnsMu.Unlock()
	}
}


